java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,354 @@
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.jglome;
+
+import static java.lang.System.arraycopy;
+import static java.security.spec.NamedParameterSpec.X25519;
+
+import java.nio.ByteBuffer;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.interfaces.XECPrivateKey;
+import java.security.interfaces.XECPublicKey;
+import java.security.spec.NamedParameterSpec;
+import java.util.Arrays;
+import java.util.Optional;
+import javax.crypto.KeyAgreement;
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * Encapsulates the logic of GLOME protocol.
+ */
+public class Glome {
+
+  private XECPublicKey peerKey;
+  private Mac userMacKey;
+  private Mac peerMacKey;
+  private int minPeerTagLength;
+
+  private final KeyPair userKeys;
+
+  private final static int MIN_CNT_VALUE = 0;
+  private final static int MAX_CNT_VALUE = 255;
+  private final static int MAX_TAG_LENGTH = 32;
+  private final static int MIN_TAG_LENGTH = 1;
+
+  /**
+   * Generates a pair of private/public keys for the user, calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum allowed peer's tag length.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  public Glome(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    this.userKeys = generateKeys();
+    initMacKeys(peerKey, minPeerTagLength);
+  }
+
+  /**
+   * Derives the user's public key from the {@code privateKey}, calculates user's and peer's MAC
+   * keys.
+   *
+   * @param peerKey peer's public key.
+   * @param userPrivateKey user's private key.
+   * @param minPeerTagLength minimum allowed peer's tag length.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  public Glome(XECPublicKey peerKey, XECPrivateKey userPrivateKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    XECPublicKey userPublicKey = derivePublicKeyFromPrivate(userPrivateKey);
+    this.userKeys = new KeyPair(userPublicKey, userPrivateKey);
+    initMacKeys(peerKey, minPeerTagLength);
+  }
+
+  public KeyPair userKeys() {
+    return userKeys;
+  }
+
+  public XECPublicKey peerKey() {
+    return peerKey;
+  }
+
+  public XECPublicKey userPublicKey() {
```

Comment:
Are both these userPublicKey/userPrivateKey and userKeys() above needed?  Seems redundant.